### ISSUE TITLE
[EPIC] Generalized command distribution

### ISSUE THREAD

**Issue by korthout** [camunda/zeebe#11456]
# Description

We know how to generalize the deployment distribution concept to use it for all record value types. The next step is to break this down into technical tasks.
- https://github.com/camunda/zeebe/issues/11218

# Open questions

- [ ] how can we deal with reaching the `maxMessageSize` twice as fast?

For deployment, the proposal requires us to write the deployed resources twice in the same record batch, i.e. once in the `Process:Created` (or `DecisionRequirements:Created`) events, and once in the `RecordDistribution:Started` event. That means that we reach the `maxMessageSize` twice as fast. In other words, users will no longer be able to deploy some of the resources they could before. **The team proposes to solve this by doubling the current `maxMessageSize`**. We will discuss this with the Zeebe Distributed Platform team, as they may challenge this.

ZDP has suggested to:
- introduce a soft limit (MAX_MESSAGE_SIZE) and a hard limit (🤷 arbitrary, perhaps half the segment size) for record batches, where some processors can choose to ignore the soft limit
- increase the limit for all record batches, and instead produce warnings when record batches exceed the MAX_MESSAGE_SIZE

Which it will be is not yet decided.

- [x] Can we treat the distributed command as a user command and acknowledge it as a response?

<details><summary>Although this question is unanswered, we've decided to move forward without using this</summary>
<p>

This question came up while deciding 'can we de-deprecate a method in the InterPartitionCommandSender?': Instead of processing distributed commands differently by adding a post-commit task to Acknowledge the command, we could treat the distributed command like other user commands and respond to it. Instead of sending the response to the gateway, it would write a command to another partition.

</p>
</details> 

- [x] can we de-deprecate [a method in the InterPartitionCommandSender](https://github.com/camunda/zeebe/blob/8.2.0-alpha3/stream-platform/src/main/java/io/camunda/zeebe/stream/api/InterPartitionCommandSender.java#L32)?

<details>
<summary>Yes, we will de-deprecate the method</summary>

<p>
Another topic that should be discussed with the Zeebe Distributed Platform (ZDP) team is de-deprecating the method to <a href="https://github.com/camunda/zeebe/blob/8.2.0-alpha3/stream-platform/src/main/java/io/camunda/zeebe/stream/api/InterPartitionCommandSender.java#L32">send a command to another partition where the sending partition determines the record key</a>. This method was deprecated immediately when it was introduced with the new InterPartitionCommandSender API to avoid us from using it for other cases than Deployment Distribution. However, we could not figure out the reason why this method should not be used. Actually, the team sees reasonable use cases where a partition can decide the record key when it sends it to another partition. As the key encodes the partition id it was generated on, it clearly identifies an entity to exist on that specific partition. We should discuss this with the Zeebe Distributed Platform team to understand why this method was deprecated in the first place and push to revise that stance.
</p>

<p>
After discussing with ZDP, we agreed that it is necessary for Deployments to re-use the key. The deployment must be available on all partitions with the same key. Otherwise, resource deletion would not be possible because we somehow need to refer to a single resource (Deployment). Initially, we thought it must also be available for process creation, but the process key is already part of the <code>ProcessMetadata</code> which is part of the <code>Deployment</code> command that is distributed to the other partitions.
</p>

<p>
Also part of the decision is that we clearly document the usage of this method. When is it allowed to use it and when not. It must also be clear why the method must exist.
</p>

<p>
For full details of the discussion, please find the associated (archived) Slack channel <a href="https://camunda.slack.com/archives/C04MCPRF5KN/p1675162591732819">#top-zeebe-dedeprecate-send-command-with-key</a>.
</p>

</details>

# Concept and goal
Some entities in Zeebe are cross-cutting partitions. For example, when you deploy a new Process, you want to be able to start Process Instances of that Process on any of the partitions. But the gateway writes the `Deployment:Create` command only on a single partition when it handles a call to the `DeployResource RPC`. The engine has to communicate this command to the other partitions (i.e. each partition has its own engine). This communication is unreliable because of the nature of distributed systems. We need a way to communicate commands between the partitions reliably. We call this mechanism command distribution.

Several features require this same mechanism. Instead of building the same logic multiple times, we should generalize the concept to reuse it for all of these. This maintenance task will unblock:
- #10777  
- #9576 
- Suspend and Resume Processes/Instances/Tasks
- ... and potentially others

**We can achieve this with the following proposed solution:**
Introduce a new value type called `CommandDistribution`, which wraps the command we want to communicate reliably to other partitions.

An initial `STARTED` intent ensures we store that command in the state (for lookups at retry), and a final `FINISHED` event can be applied to remove this command from the state.

Then, for every partition, we have intents `DISTRIBUTING`, `ACKKNOWLEDGE`, and `ACKNOWLEDGED` to keep track of the distribution to that specific partition.

On the receiving partition, we write the encapsulated command directly. Its key contains the information that it was distributed (and from which partition), so the engine knows the send the `ACKNOWLEDGE` back to the distributing partition.

# Task breakdown

```[tasklist]
### Tasks
- [x] #11657
- [x] https://github.com/camunda/zeebe/issues/11868
- [x] #11658
- [x] #11659
- [x] #11660
- [ ] https://github.com/camunda/zeebe/issues/12764
- [ ] #11914
- [ ] #11661
- [ ] https://github.com/camunda/zeebe/issues/13057
- [ ] https://github.com/camunda/zeebe/issues/13058
- [ ] https://github.com/camunda/zeebe/issues/13164
- [ ] https://github.com/camunda/zeebe/issues/13161
- [ ] https://github.com/camunda/zeebe/issues/13162
- [ ] https://github.com/camunda/zeebe/issues/13163
- [ ] https://github.com/camunda/zeebe/issues/13233
```

**Comment by korthout** [comment]
Moving this to `blocked` while deciding on the open questions. These are expected to be answered at the start of the upcoming week.

**Comment by korthout** [comment]
Moved this into `in progress`. Open questions have either been answered or don't block progress.

**Comment by korthout** [comment]
Upgrading this issue to `EPIC` as the breakdown is now clear, and it would be silly to copy the open questions, concept, and tasks breakdown into a new issue just for that.

**Comment by remcowesterhoud** [comment]
@korthout do we want to close this epic and leave the remaining issue as a standalone improvement?

**Comment by korthout** [comment]
I'm investigating a small regression (max deployable resource payload reduced from ~2MB to ~1.4MB on multi-partition clusters) due to
- #11661 

We can keep it separate but might be good just to track it here until completed.

**Comment by korthout** [comment]
We can track #13233 separately. Let's close this epic! 🎉
